core/src/main/scala/org/apache/spark/CacheRecoveryManager.scala

```diff
@@ -0,0 +1,189 @@
+
+    val (expiringMemStatus, remainingMemStatus) = allExecMemStatus.partition {
+      case (execId, _) => execsToShutDown.contains(execId)
+    }
+    val freeMemOnRemaining = remainingMemStatus.values.map(_._2).sum
+
+    // The used mem on each executor sorted from least used mem to greatest
+    val executorAndUsedMem: Seq[(String, Long)] =
+      expiringMemStatus.map { case (execId, (maxMem, remainingMem)) =>
+        val usedMem = maxMem - remainingMem
+        execId -> usedMem
+      }.toSeq.sortBy { case (_, usedMem) => usedMem }
+
+    executorAndUsedMem
+      .scan(("start", freeMemOnRemaining)) {
+        case ((_, freeMem), (execId, usedMem)) => (execId, freeMem - usedMem)
+      }
+      .drop(1)
+      .filter { case (_, freeMem) => freeMem > 0 }
+      .map(_._1)
+  }
+
+  /**
+   * Given an executor id, start a timer that will kill the given executor after the configured
+   * timeout
+   *
+   * @param execId The id of the executor to be killed
+   * @return a future representing the timer
+   */
+  private def startKillTimer(execId: String): ScheduledFuture[_] = {
+    val killer = new Runnable {
+      def run(): Unit = {
+        logDebug(s"Killing $execId because timeout for recovering cached data has expired")
+        kill(execId)
+      }
+    }
+    scheduler.schedule(killer, forceKillAfterS, TimeUnit.SECONDS)
+  }
+
+  /**
+   * Recover cached RDD blocks off of an executor until there are no more, or until
+   * there is an error
+   *
+   * @param execId the id of the executor to be killed
+   * @param killTimer The runnable scheduled to kill this executor. Cancel it if we finish before
+   *                  it does.
+   */
+  private def replicateUntilDone(execId: String, killTimer: ScheduledFuture[_]): Unit = {
+    recoverLatestBlock(execId).onComplete {
+      case Succ(Some(blockId)) =>
+        logTrace(s"Replicated block $blockId on executor $execId")
+        replicateUntilDone(execId, killTimer)
+      case Succ(None) =>
+        killTimer.cancel(false)
+        kill(execId)
+      case Failure(e) =>
+        logWarning("Failure recovering cached data before executor $execId shutdown", e)
+        killTimer.cancel(false)
+        kill(execId)
+    }
+  }
+
+  /**
+   * Replicate the latest cached rdd block off of this executor on to a surviving executor, and then
+   * remove the block from this executor
+   *
+   * @param execId the executor to recover a block from
+   * @return A future holding the id of the block that was recovered or None if there were no blocks
+   *         to recover.
+   */
+  private def recoverLatestBlock(execId: String): Future[Option[RDDBlockId]] = {
+    blockManagerMasterEndpoint
+      .ask[Option[RDDBlockId]](ReplicateLatestRDDBlock(execId, recoveringExecutors.toSeq))
+      .andThen {
+        case Succ(Some(blockId)) =>
+          blockManagerMasterEndpoint.ask[Boolean](RemoveBlockFromExecutor(execId, blockId))
```

Comment:
I think this would be fine, the block removal would always happen after the replicate was done, but I think I'm just going to merge those two messages based on your later comment.